test/Streamly/Test/FileSystem/Handle.hs

```diff
@@ -0,0 +1,153 @@
+-- |
+-- Module      : Streamly.Test.FileSystem.Handle
+-- Copyright   : (c) 2020 Composewell technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+
+module Streamly.Test.FileSystem.Handle (main) where
+
+import Data.Functor.Identity (runIdentity)
+import Data.Word (Word8)
+import Foreign.Storable (Storable(..))
+import System.IO
+    ( Handle
+    , IOMode(..)
+    , SeekMode(..)
+    , hFlush
+    , hSeek
+    , openFile
+    )
+import System.IO.Temp (writeSystemTempFile)
+import Test.QuickCheck (Property, forAll, Gen, vectorOf, choose)
+import Test.QuickCheck.Monadic (monadicIO, assert, run)
+
+import Streamly.FileSystem.Handle as FH
+import Streamly.Internal.Data.Stream.IsStream (IsStream, SerialT)
+import Test.Hspec as H
+import Test.Hspec.QuickCheck
+
+import qualified Streamly.Data.Fold as FL
+import qualified Streamly.Internal.Data.Stream.IsStream as S
+import qualified Streamly.Internal.Data.Array.Storable.Foreign as A
+import qualified Streamly.Internal.Unicode.Stream as U
+
+allocOverhead :: Int
+allocOverhead = 2 * sizeOf (undefined :: Int)
+
+defaultChunkSize :: Int
+defaultChunkSize = 32 * k - allocOverhead
+   where k = 1024
+
+maxArrLen :: Int
+maxArrLen = defaultChunkSize * 8
+
+maxTestCount :: Int
+maxTestCount = 10
+
+chooseWord8 :: (Word8, Word8) -> Gen Word8
+chooseWord8 = choose
+
+utf8ToString :: A.Array Word8 -> String
+utf8ToString = runIdentity . S.toList . U.decodeUtf8' . A.toStream
+
+testData :: String
+testData = "This is the test data for FileSystem.Handle ??`!@#$%^&*~~))`]"
+
+testDataLarge :: String
+testDataLarge = concat $ replicate 6000 testData
+
+tempHandle :: IO Handle
+tempHandle = do
+    fp <- writeSystemTempFile "fs_handle" testDataLarge
+    openFile fp ReadMode
+
+tempHandleW :: IO Handle
+tempHandleW = do
+    fp <- writeSystemTempFile "fs_handle" ""
+    openFile fp ReadWriteMode
+
+readFromHandle :: IO (SerialT IO Char)
+readFromHandle = do
+    h <- tempHandle
+    return $ U.decodeUtf8 $ S.unfold FH.read h
```

Comment:
We are not closing the handles. It should not be a problem if we are running the test only a few times, but if we run it many times we may run out of handles.